dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSBenchmarkTest.java

```diff
@@ -0,0 +1,366 @@
+            for (int j = 0; j < numReadSamples; j++)
+            {
+                long startTime = System.currentTimeMillis();
+                try
+                {
+                    readFileSerially(file);
+                }
+                catch(Exception e)
+                {
+                    Assert.fail(e.getMessage());
+                }
+
+                long endTime = System.currentTimeMillis();
+
+                float timeInSeconds = (endTime - startTime);
+                timeInSeconds /= 1000.0;
+                serialReadTimes.add(new Float(timeInSeconds));
+            }
+
+            // Remove longest sample (Remove long samples due to wwarmup)
+            longestSample = serialReadTimes.get(0);
+            sampleToRemove = 0;
+            for (int j = 0; j < serialReadTimes.size(); j++) {
+                Float currentReadSample = serialReadTimes.get(j);
+                if (longestSample < currentReadSample) {
+                    longestSample = currentReadSample;
+                    sampleToRemove = j;
+                }
+            }
+            serialReadTimes.remove(sampleToRemove);
+
+            float avgSerialReadTime = 0;
+            for (int j = 0; j < serialReadTimes.size(); j++) {
+                avgSerialReadTime += serialReadTimes.get(j);
+            }
+            avgSerialReadTime /= serialReadTimes.size();
+
+            float recordConstructionEfficiency = avgRawReadTime / avgSerialReadTime;
+            if (recordConstructionEfficiency < minDatasetConstructionEfficiency[i]) {
+                Assert.fail("Record construction efficiency lower than expected for dataset: " 
+                    + datasets[i] + " min efficiency: " + minDatasetConstructionEfficiency[i]
+                    + " benchmarked efficiency: " + recordConstructionEfficiency);
+            }
+            System.out.println("Record construction efficiency: " + recordConstructionEfficiency + " for: " + datasets[i]);
+        }
+    }
+
+    public void readRawFileData(HPCCFile file) throws Exception
+    {
+        if (file == null)
+        {
+            Assert.fail("HPCCFile construction failed.");
+        }
+
+        DataPartition[] fileParts = file.getFileParts();
+        if (fileParts == null || fileParts.length == 0)
+        {
+            Assert.fail("No file parts found");
+        }
+
+        FieldDef originalRD = file.getRecordDefinition();
+        if (originalRD == null || originalRD.getNumDefs() == 0)
+        {
+            Assert.fail("Invalid or null record definition");
+        }
+
+        byte[] buffer = new byte[4 * 1024 * 1024];
+        for (int i = 0; i < fileParts.length; i++)
+        {
+            RowServiceInputStream inputStream = new RowServiceInputStream(fileParts[i],originalRD,originalRD,120,-1);
+
+            boolean hasMoreData = inputStream.available() > 0;
+            if (hasMoreData == false) {
```

Comment:
are we trying to read from inputstream when  inputstream.available <= 0?

Reply:
Yep. Available is the amount we can read right now not the amount left in the stream. So to check for the end of the stream we try and read and if they read fails we have reached the end of the stream.